File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
public:
    stack<int> mainStack;
    stack<int> minStack; 
    MinStack() {
        
    }
    
    void push(int val) {
        mainStack.push(val); 
        if(minStack.empty() || val <= minStack.top()) minStack.push(val); 
    }
    
    void pop() {
        int top_ele = top(); 
        if(!minStack.empty() && minStack.top() == top_ele) minStack.pop(); 
        mainStack.pop(); 
    }
    
    int top() {
        return mainStack.top(); 
    }
    
    int getMin() {
        return minStack.top(); 
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155-min-stack/0155-min-stack.java
class MinStack {

    Deque<Integer> mainStack = new ArrayDeque<>(); 
    Deque<Integer> minStack = new ArrayDeque<>();

    public MinStack() {
        
    }
    
    public void push(int value) {
        mainStack.addLast(value); 
        if(minStack.isEmpty() || minStack.peekLast() >= value){
            minStack.addLast(value); 
        }
    }
    
    public void pop() {
        int target = mainStack.pollLast(); 
        if(!minStack.isEmpty() && minStack.peekLast() == target) minStack.pollLast(); 
    }
    
    public int top() {
        return mainStack.peekLast(); 
    }
    
    public int getMin() {
        return minStack.peekLast(); 
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(value);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */